hive-bigquery-connector-common/src/main/java/com/google/cloud/hive/bigquery/connector/BigQueryStorageHandlerBase.java

```diff
@@ -257,19 +253,62 @@ public void configureOutputJobProperties(TableDesc tableDesc, Map<String, String
     JobDetails jobDetails = new JobDetails();
     HiveBigQueryConfig opts = HiveBigQueryConfig.from(conf, tableProperties);
     jobDetails.setWriteMethod(opts.getWriteMethod());
-    jobDetails.setBigquerySchema(bqTableInfo.getDefinition().getSchema());
+    jobDetails.setBigquerySchema(bigQuerySchema);
     jobDetails.setTableProperties(tableProperties);
-    jobDetails.setTableId(tableId);
+    jobDetails.setTableId(
+        BigQueryUtil.parseTableId(tableProperties.getProperty(HiveBigQueryConfig.TABLE_KEY)));
 
     if (opts.getWriteMethod().equals(HiveBigQueryConfig.WRITE_METHOD_INDIRECT)) {
+      Injector injector =
+          Guice.createInjector(
+              new BigQueryClientModule(), new HiveBigQueryConnectorModule(conf, tableProperties));
       configureJobDetailsForIndirectWrite(
           opts, jobDetails, injector.getInstance(BigQueryCredentialsSupplier.class));
     }
 
+    try {
+      createBigQueryTableIfDoesNotExist(jobDetails);
+    } catch (Exception e) {
+      throw new RuntimeException(e + ". The KMS key is: " + opts.getKmsKeyName());
+    }
+
     // Save the job details file to disk
     jobDetails.writeFile(conf);
   }
 
+  /**
+   * This function determines whether the destination table exists: if it doesn't, then create it.
+   */
+  public void createBigQueryTableIfDoesNotExist(JobDetails jobDetails)
+      throws IllegalArgumentException {
+    Injector injector =
+        Guice.createInjector(new BigQueryClientModule(), new HiveBigQueryConnectorModule(conf));
+    BigQueryClient bqClient = injector.getInstance(BigQueryClient.class);
+    HiveBigQueryConfig opts = injector.getInstance(HiveBigQueryConfig.class);
+    if (bqClient.tableExists(jobDetails.getTableId())) {
+      // Check that the destination table's schema matches that of the insert query
+      TableInfo destinationTable = bqClient.getTable(jobDetails.getTableId());
+      Schema destinationTableSchema = destinationTable.getDefinition().getSchema();
+      Preconditions.checkArgument(
+          BigQueryUtil.schemaWritable(
+              jobDetails.getBigquerySchema(),
+              destinationTableSchema,
+              false, /* regardFieldOrder */
+              opts.getEnableModeCheckForSchemaFields()),
+          new InvalidSchemaException(
+              "Destination table's schema is not compatible with query's schema"));
+      jobDetails.setDeleteTableOnAbort(false);
```

Comment:
The idea here is that if the BQ table doesn't exist, we assume that the job itself will create it before inserting new rows. So, if the job fails, we want to clean it up on exit. If the table already exists, then we want to leave it alone.

`JobDetails.deleteTableOnAbort` already defaults to false, so that statement is in fact not strictly needed. I just have it there for clarity. So I don't think the scenario you mention could in fact happen.

What do you think?